Feat/filter buttons

jcee-io opened the pull request:
Read previously closed PR with the same title as this one to view other changes.

Additional changes:

**Tests added**
- For single filters, double filters, and all filters combined
- Tests are passing

**UX changes**
- Thanks to my internet being slow upstairs there was a small bug/not bug where clicking the filter button BEFORE parklets were rendered would not do anything since the data automatically gets set as the state for activeFilter (aka I'm using a refresh method with an empty array if I click a filter button). This has been fixed by calling the refresh method right after data has been loaded through componentDidMount

jcee-io committed:
(feat) create FilterButton jsx file with skeleton included

jcee-io committed:
(feat) includes className for original button design

jcee-io committed:
(feat) implement toggle on filter button

jcee-io committed:
(refactor) remove hover animations

jcee-io committed:
(refactor) use FilterButton for functionality instead of vanilla html

jcee-io committed:
(bug) <i> element click changes button color now instead of just icon

jcee-io committed:
(feat) change App to be stateful and create a function that handles filters

jcee-io committed:
(refactor) parklets load through App component

jcee-io committed:
(feat) implement handleFilters function to respond to filter button clicks

jcee-io committed:
(feat) implement data faker

jcee-io committed:
(refactor) implant fake data methods into the promise block

jcee-io committed:
(refactor) clean up for linter and add error state to import from MapView

jcee-io committed:
(feat) implement prop types for components that now use props

jcee-io committed:
(refactor) change error origin from MapView to App

jcee-io committed:
(refactor) fix 'Enter Address' from being a FilterButton to regular button

jcee-io committed:
(refactor) adhere compiled SCSS to linter guidelines

jcee-io committed:
(refactor) switch helper function to a standalone js file

jcee-io committed:
(refactor) use state for parklet filters instead of using dom selectors

jcee-io committed:
(refactor) create helper function to refresh parklet display from handleFilters

jcee-io committed:
(refactor) move fakeFilterFlags into util folder and use ES6 syntax

jcee-io committed:
(refactor) import fakeFilterFlags from util folder

jcee-io committed:
(refactor) use filter states from App instead of DOM manipulation

jcee-io committed:
(refactor) change filter flags to not be random

jcee-io committed:
(refactor) leave one parklet with all flags true

jcee-io committed:
(refactor)add  default propTypes

jcee-io committed:
(refactor) switch parklet data fetching to App instead of MapView

jcee-io committed:
(feat) add tests for FilterButton functionality with response to parklet data

jcee-io committed:
(refactor) remove FilterButton functionality tests from App spec file

jcee-io committed:
(refactor) render parklet correctly when clicking filter button before data is fetched

jcee-io committed:
(refactor) reduce amount of setState changes to 1 in componentDidMount without removing the UX changes

andino93 commented on client/src/App.jsx:
```diff
@@ -1,13 +1,86 @@
+    const activeParklets = [];
+    const hiddenParklets = [];
+
+    parklets.forEach((parklet) => {
```
I would use reduce instead to sort parklets into an object with your two keys. Then you can do setState with `{...filteredObject}`.

andino93 commented on client/src/FilterButton.jsx:
```diff
@@ -0,0 +1,56 @@
+import React from 'react';
+import PropTypes from 'prop-types';
+
+const getType = (type) => {
```
filterButton should only contain the component

jcee-io replied:
For both functions that need to be moved out, should I put it both on utils, or do I move them over to App?

andino93 commented on client/src/FilterButton.jsx:
```diff
@@ -0,0 +1,56 @@
+  }
+};
+
+const toggleFilter = (event, handler, type) => {
```
filterButton should only contain the component

andino93 commented on client/src/FilterButton.jsx:
```diff
@@ -0,0 +1,56 @@
+  }
+};
+
+const FilterButton = props => {
```
missing parens around props. check your linter

andino93 commented on client/src/App.jsx:
```diff
@@ -1,13 +1,86 @@
+    // IMPORTANT: Until db is implemented, fake data will be used to generate flags
+    // for a parklet that is open, has coffee, or wifi by RNG, this will change for
+    // every refresh of the page
+    fetchParklets(PARKLETS_ENDPOINT)
```
Can you change this to just load the fake data instead of regenerating it every time the component mounts. This way, you can structure the component in a way to replace the fake data with real data without change in codebase.

jcee-io committed:
(refactor) add conditional to prevent fetching with existing data

jcee-io committed:
(refactor) change parklet data handling from forEach to reduce

jcee-io committed:
(refactor) move filter button checker to utils, and removed direct dom access from FilterButtons

jcee-io commented:
Changes addressed on @andino93's review:
**App**
- componentDidMount now only fetches parklet data if and only if activeParklet and hiddenParklet states are both completely empty
- Parklet refreshing now uses reduce instead of forEach for a cleaner syntax (minor change)

**FilterButton**
- getType is now in utils
- filterClass is now in utils 
- DOM manipulation removed and filter button class now based on filter states from App
- all FilterButton utils located in utils/getFilterButtonType.jsx

mikehern commented on spec/client/FilterButtons.spec.jsx:
```diff
@@ -0,0 +1,85 @@
+    const { activeParklets, hiddenParklets } = wrapper.state();
+    const wifiOnly = [...activeParklets, ...hiddenParklets].filter(parklet => parklet.wifi);
+
+    expect(activeParklets).toHaveLength(wifiOnly.length);
```
It looks like your unit tests are close to testing your implementation detail.

Let me know if I'm misunderstanding your fakeFilterFlags helper; it looks like it takes the entire set of 31 parklets, sets the first third to be "closed", the second third to not have wifi, and the last 2/3rds to not have food.

The expect statement in each test asserts the number of filtered parklets in the state is the same value as the length of the filtered dataset.

I realize it is subtle because your implementation pushes parklets into arrays, while the tests use filter.  So one could argue that `Array.push` and `Array.filter` are different "implementations".

If the helper applies filters to a fixed/constant number of parklets, why not just hardcode your tests to expect that the state match those fixed values... 1/3 * 31, 2/3 *31, etc?

jcee-io replied:
I actually forgot about that portion of the test, at the time of writing it I did not want to count nor rationalize how many parklets are there in an active filter button, so I used a filter function to filter the same amount of parklet as it does during the state since I'm also working on multiple active filter button as well. I should probably change it to hardcode in that case.  What are your thoughts on this? (I actually might change it to hardcode if I have time before I sleep)

mikehern commented on client/utils/fakeFilterFlags.js:
```diff
@@ -0,0 +1,29 @@
+    node.open = true;
+  });
+
+  data[0].wifi = true;
```
No need for any changes here, but I am curious as to why the first parklet in the array is manually set to these values after setting all of the values already in the 3 slice functions?

jcee-io replied:
Oh I'm glad you asked that, because I'm fixing the flags until a more legitimate method works. This is in order to have at least one parklet that has all three flags to display the behavior when all three filters are active.

mikehern reviewed (commented):
1 question about unit testing, and 1 question about the data decorator helper.

jcee-io committed:
(refactor) hard code expected length for parklets

jcee-io commented:
**Test changes from @mikehern's suggestion**
- Expected length is now a hardcoded number instead of a filtered array's length
- The one-button filters are 21, 21, 23 in length for wifi, food, and open. The reason for this is because of the respective parklet's indexes based off of Math.floor(), check that file out for more details. Because the amount of parklets isn't a multiple of 3, it does give a weird effect on the array lengths of having two active filters. Nonetheless, all tests are being passed due to the constant nature of fakeFilterFlags.js applying to the same indexes for every fetch.

**Addressing Concerns**
- The hardcoded numbers are based on the length of the filtered array when checking for flags, and not the activeParklets state. I can see where room for biased testing might occur with this change. 
- Should the previous version of using filtered arrays be more sufficient, it's only one commit I can easily roll back, so let me know if this is preferred.

jcee-io committed:
(feat) add basic route for fetching parklet in server

jcee-io committed:
(feat) wire up server to do data fetching and return to client instead of raw client fetching

jcee-io committed:
(refactor) minimize use of fake data methods on App component and focuses on fetchParklets

jcee-io committed:
(feat) install axios to make http requests on node

jcee-io commented:
**Update**
- Simply move calling SFgov with parklet data from the client to the server.
- I managed to get the tests to pass easily from changing the SFgov URL to '/parklet'
- App component does not use helper functions such as fakeFilterFlags, this has been migrated to fetchParklets

**Some Questionable Things That Will Get Pointed Out Anyways**
- It's intentional that I left fetchParklets having to format the json (AKA "hacking it" - @mikehern ). It behaves just the same when calling these functions on the server. However, for some odd reason, it fails the test "corrects the coordinates from lon-lat to lat-lon" for mapView.spec.jsx having the state being the raw, unmodified SFgov data without the use of the helper functions intended to give it the { title, position } format. I speculate this might be because of the parklets.json file in the clients folder, but the FilterButtons test suite passes.
 **TL;DR Intended behavior on GET requesting the json or the UI, tests were rekt

- Piggybacking off of that, the said corrections of the lon-lat format on mapView and the helper functions needed to "hack" said fake data would be obsolete by the time we have a working database. I would like to open up discussion on whether or not it would be worth stressing more and pursuing further changes.
- Another issue is using fakeFilterFlags on server, it messes up FilterButtons test suite. 

** SUMMARY: Unless I use helper functions on the client-side, it seems parklets.json and fetchMock give back the format pre-modified even if I res.json() a modified json through. Again the behavior of the actual client is the same either way, it's just the test suite acting weird.

**What Can Be Improved**
- Find a way to move those helper functions into server without failing the tests perhaps? I'm aware it contradicts the last section.
 **If we pursue using this PR anyways,  I would need some help trying to figure out how to improve the test suite

**Miscellaneous**
- So this branch and PR is intended for simply the functionality of the filter buttons with respect to filtering parklets and all tests showcasing functionality. I suggest future improvements from this PR be directed to a new branch post-merge.  The PR itself is large enough, and it would stray from what this was originally intended. Feel free to let me know any thoughts on that (I'm more than happy to make any more improvements in this same PR if we have a consensus on that). 

P.S. It's 2 am and I'm pretty tired, so my apologies for the wall of text and future commits after this comment, it's mostly going to be syntactical clean-ups to make the linter happy. If you read through the whole thing, you have my gratitude and deserve a high five.

jcee-io committed:
(refactor) handle fetching errors through json responses

andino93 commented on client/src/App.jsx:
```diff
@@ -1,13 +1,96 @@
+      fetchParklets(PARKLETS_ENDPOINT)
+        .then((data) => {
+          if (data.error) {
+            throw new Error(data.error);
```
new keyword unncessary

andino93 replied:
Is there also a reason you're throwing an error here but then setting an error in state on line 38? Throwing errors is generally not a great way of handling errors.

andino93 commented on spec/client/App.spec.jsx:
```diff
@@ -1,8 +1,13 @@
 import React from 'react';
 import renderer from 'react-test-renderer';
+import { shallow } from 'enzyme';
```
Is there a reason new modules and files were added to this file and then not used?

jcee-io replied:
Nope, just accidental leftovers. Probably from moving tests to FilterButton spec files IIRC. Just updated it to remove those modules.

jcee-io committed:
(refactor) remove imported unused modules on spec files

mikehern commented on client/src/Interface.jsx:
```diff
@@ -1,11 +1,13 @@
+import FilterButton from './FilterButton';
 
-const Interface = () => (
+const Interface = props => (
```
I think our linter should have caught this?

https://github.com/airbnb/javascript#arrows--confusing

Edit: nevermind, it's not conditional.

jcee-io committed:
(refactor) remove error throwing in App component and use conditional instead
